File: GolangSolutions/String/M1790_仅执行一次字符串交换能否使两个字符串相等.go
package String

func areAlmostEqual(s1 string, s2 string) bool {
	if s1 == s2 {
		return true
	}
	n := len(s1)
	t1, t2 := uint8(' '), uint8(' ')
	flag := false
	for i := 0; i < n; i++ {
		if s1[i] != s2[i] {
			if t1 == ' ' {
				t1 = s1[i]
				t2 = s2[i]
			} else if !flag && t1 == s2[i] && t2 == s1[i] {
				flag = true
			} else {
				return false
			}
		}
	}

	return flag
}

func areAlmostEqual2(s1 string, s2 string) bool {
	i, j := -1, -1
	for idx := range s1 {
		if s1[idx] != s2[idx] {
			if i < 0 {
				i = idx
			} else if j < 0 {
				j = idx
			} else {
				return false
			}
		}
	}

	return i < 0 || j >= 0 && s1[i] == s2[j] && s1[j] == s2[i]
}

File: CPPSolutions/CPPSolutions/M1790_仅执行一次字符串交换能否使两个字符串相等.cpp
#include <cstdio>
#include <string>

using namespace std;

class Solution {
public:
    bool areAlmostEqual(string s1, string s2) {
        if (s1 == s2) return true;
        int n = s1.size();
        char t1 = ' ', t2 = ' ';
        bool flag = false;
        for (int i = 0; i < n; i++) {
            if (s1[i] != s2[i])
                if (t1 == ' ')
                    t1 = s1[i], t2 = s2[i];
            else if (!flag && t1 == s2[i] && t2 == s1[i])
                flag = true;
            else 
                return false;
        }

        return flag;
    }

    bool areAlmostEqual2(string s1, string s2) {
        int n = s1.size();
        int i = -1, j = -1;
        for (int idx = 0; idx < n; idx++) {
            if (s1[idx] != s2[idx]) {
                if (i < 0)
                    i = idx;
                else if (j < 0)
                    j = idx;
                else 
                    return false;
            }
        }

        return i < 0 || j >= 0 && s1[i] == s2[j] && s1[j] == s2[i];
    }
};